botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1123 @@
+    EndpointInputParametersError,
+    EndpointResolutionError,
+    RuleSetDefinitionError,
+)
+from botocore.utils import (
+    ArnParser,
+    InvalidArnException,
+    is_valid_ipv4_endpoint_url,
+    is_valid_ipv6_endpoint_url,
+    normalize_url_path,
+    percent_encode,
+)
+
+logger = logging.getLogger(__name__)
+
+TEMPLATE_STRING_RE = re.compile(r"\{[a-zA-Z#]+\}")
+VALID_HOST_LABEL_RE = re.compile(
+    r"^((?!-)[a-zA-Z\d-]{1,63}(?<!-))*((?!-)[a-zA-Z\d-]{1,63}(?<!-))$",
+)
+GET_ATTR_RE = re.compile(r"(\w+)\[(\d+)\]")
+CACHE_SIZE = 100
+
+
+class BaseRule:
+    """A rule within a rule set. All rules contain a conditions property,
+    which can be empty, a partitions JSON and documentation about the rule.
+    """
+
+    def __init__(self, conditions, partitions, documentation=None):
+
+        self.conditions = conditions
+        self.partitions = partitions
+        self.documentation = documentation
+
+    def evaluate(self, scope_vars):
+        raise NotImplementedError()
+
+    def is_func(self, argument):
+        """Determine if an object is a function object.
+
+        :type argument: Any
+        :param argument: An object that may or may not be a function object.
+
+        :returns: A bool indicating if the input is a function object.
+        """
+        return isinstance(argument, dict) and "fn" in argument
+
+    def is_ref(self, argument):
+        """Determine if an object is a reference object.
+
+        :type argument: Any
+        :param argument: An object that may or may not be a reference object.
+
+        :returns: A bool indicating if the input is a reference object.
+        """
+        return isinstance(argument, dict) and "ref" in argument
+
+    def is_template(self, argument):
+        """Determine if an object contains a template string.
+
+        :type argument: Any
+        :param argument: An object that may or may not contain a template string.
+
+        :returns: A bool indicating if the input contains a template string.
+        """
+        return (
+            isinstance(argument, str)
+            and TEMPLATE_STRING_RE.search(argument) is not None
+        )
+
+    def resolve_template_string(self, value, scope_vars):
+        """Given a templated string, parse out the template nested between curly
+        braces, and return the nested data from the rule's scoped variables. If the
+        string contains double curly braces such as `{{foo}}` they are treated as
+        the string literal `{foo}` and are not formatted.
+
+        :type value: str
+        :param value: A template string such as "https://{Bucket}.ec2.{url#authority}".
+
+        :type scope_vars: dict
+        :param scope_vars: A dictionary combining input parameters and variables that
+        have been scoped throughout the rule's evaluation.
+
+        :returns: A fully resolved string such as "https://us-west-2.ec2.amazonaws.com".
+        """
+        string_formatter = Formatter()
```

Comment:
This likely doesn't need to be created for every call.

Reply:
its been moved to a module constant